### PR TITLE
Issue 48968 and 48959

### PR DESCRIPTION
**Pull request by XingY**
Issue [48968](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=48968): LKSM: Updating samples with sample IDs that look like integers are getting confused
Issue [48959](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=48959): Sample Manager: Ancestor naming pattern is not respecting specified parent

#### Related Pull Requests
* 

#### Changes
*

### CODE REVIEW

File: api/src/org/labkey/api/data/NameGenerator.java [hunk 1]
```diff
@@ -1256,9 +1256,9 @@ private List<String> processFieldParts(FieldKey fkTok, Map<String, NameExpressio
                     return fieldParts;
                 }
 
-                if (ancestorLevel > 4) // 1 generation of direct parent + 4 extra generations of ancestors
+                if (ancestorLevel > 9) // 1 generation of direct parent + 9 extra generations of ancestors
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/data/NameGenerator.java:1259]: The 9 here is somewhat arbitrary, right?  Might be worth a comment somewhere.

File: api/src/org/labkey/api/data/NameGenerator.java [hunk 2]
```diff
@@ -1271,7 +1271,7 @@ private List<String> processFieldParts(FieldKey fkTok, Map<String, NameExpressio
 
         if (!ancestorTypes.isEmpty())
         {
-            if (!hasLookupColumn)
+            if (!hasLookupColumn || fieldParts.size() < 2 || fieldParts.size() > 4)
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/data/NameGenerator.java]: Might want a comment here about the size limit as well and how it relates to the ancestorLevel.

File: api/src/org/labkey/api/data/NameGenerator.java [hunk 3]
```diff
@@ -1283,7 +1283,24 @@ private List<String> processFieldParts(FieldKey fkTok, Map<String, NameExpressio
...
+            String parentTypeName = null;
+            if (parentParts.size() == 1) // alias, or one of Inputs/, MaterialInputs/, DataInputs/
+            {
+                String alias = parentParts.get(0);
+                boolean isParentAlias = importAliases != null && importAliases.containsKey(alias);
+                if (isParentAlias)
+                {
+                    String dataTypeToken = importAliases.get(alias);
+                    parentTypeName = dataTypeToken.substring(dataTypeToken.indexOf("/") + 1);
+                }
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/data/NameGenerator.java:1297]: if `isParentAlias` is false (meaning we have something like "MaterialInputs/Blood"), `parentTypeName` will still be null, right?  Is that what we want?

**Review by labkey-susanh — APPROVED** [review]

File: api/src/org/labkey/api/data/NameGenerator.java (hunk 3 above)
Comment (XingY) [reply api/src/org/labkey/api/data/NameGenerator.java:1297]: If not alias, then it's one of those general Inputs/, MaterialInputs/, DataInputs/, in which case we want to parentTypeName to be null to accept all types.

File: api/src/org/labkey/api/data/NameGenerator.java (hunk 1 above)
Comment (XingY) [reply api/src/org/labkey/api/data/NameGenerator.java:1259]: it is in fact arbitrary... someone wanted 6, so Hannah decided to make it 10 lol

**Pull request merged**